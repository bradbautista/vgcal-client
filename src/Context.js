import React from 'react'

const Context = React.createContext({
    releases: [],    
})

export default Context