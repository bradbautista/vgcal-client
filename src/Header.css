.header {
    background-color: blue;
    height: auto;
    padding: 20px;
    display: flex;
    flex-direction: column;
    align-items: flex-start;
    justify-content: space-between;
    font-size: 12px;
    letter-spacing: 0px;
    font-family: 'Press Start 2P';
    color: white;
}

.logotype {
    flex-grow: 1;
    margin-bottom: 15px;
}

.search {
    border-radius: 5px;
    border: none;
    padding: 0px;
    width: 100%;
    color: black;
    font-family: Consolas;
    font-size: 16px;
    z-index: 2;
}

@media screen and (min-width: 1200px) {

    .header {
        flex-direction: row;
        align-items: center;
        height: 75px;
    }
    
    .logotype {
        flex-grow: 0;
        margin-bottom: 0;
    }
    
    .search {
        border-radius: 5px;
        border: none;
        padding: 0px;
        width: 35%;
    }

}