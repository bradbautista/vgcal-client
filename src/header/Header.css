.header {
    background-color: var(--primary);
    height: auto;
    padding: 20px;
    display: flex;
    flex-direction: row;
    align-items: flex-start;
    justify-content: space-between;
    font-size: 16px;
    letter-spacing: 0px;
    font-family: 'Roboto Slab', serif;
    color: white;
    flex-wrap: wrap;
}

.logotype {
    margin-bottom: 15px;
    margin-top: -5px;
    padding: 0;
    order: 1;
    font-weight: 700;
}

.search {
    border-radius: 5px;
    border: none;
    padding: 0px;
    width: 100%;
    color: black;
    font-family: 'Roboto Slab', serif;
    letter-spacing: 0.35px;
    line-height: 14px;
    font-size: 14px;
    order: 3;
    z-index: 2;
    margin-top: 6px;
}

.header-nav {
    display: flex;
    flex-direction: row;
    align-items: center;
    order: 2;
}

i {
    font-family: FontAwesome;
    font-size: 32px;
    justify-self: center;
    cursor: pointer;
}

.fa-github {
    margin-left: 25px;
    margin-right: 25px;
}

@media screen and (min-width: 1200px) {

    .header {
        flex-direction: row;
        align-items: flex-start;
        height: 75px;
        /* box-shadow: 0px 3px 20px 0px rgba(50, 50, 50, 0.66); */
        /* box-shadow: 0px 0.25px 4px 0px white; */
            
    }
    
    .logotype {
        flex-grow: 0;
        padding: 0px;
        margin-bottom: 0px;
        margin-left: 10px;
        /* margin-top: -10px; */
        vertical-align: top;
    }
    
    .search {
        border-radius: 5px;
        border: none;
        padding: 0px;
        width: 35%;
        order: 2;
        margin-top: 0px;
    }

    .fa-github {
        margin-right: 0px;
    }

    .header-nav {
        order: 3;
        margin-right: 20px;
    }

    i {
        margin-left: 28px;
    }

}

@media screen and (min-width: 1400px) {
    .header-nav {
        margin-right: 40px;
    }
}