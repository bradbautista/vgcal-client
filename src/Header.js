import React, { Component } from 'react'
import { Link } from "react-router-dom";
import Select from 'react-select';
import Context from './Context';
import './Header.css'

  export default class Header extends Component {
  
    constructor(props) {

      super(props);
      this.state = {
      };
  
    }

    static contextType = Context; 
  
    render() {
  
      return (
        <header className="header">
          {/* Onclick here was this.context.setLastVisited('/')  */}
          <Link to="/" onClick={() => window.location.reload()}><h1 className="logotype">vgCal</h1></Link>
          <Select
            className="search"
            onChange={this.context.handleSelect}
            placeholder="Select a game or type to search for one"
            options={this.context.games.map(game => {
              
              const tempObj = {}

              tempObj.value = {
                platforms: game.platforms,
                image: game.boxart_url,
                gameUrl: `/game/${game.game_name.split(' ').join('-')}`,
                gameTitle: game.game_name,
                releaseDate: game.release_date_UTC,
                description: game.game_description
              };

              tempObj.label = game.game_name;

              return tempObj;

            })}
          />
          <span></span>
        </header>
      );
    }
  }