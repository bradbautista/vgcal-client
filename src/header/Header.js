import React, { Component } from 'react'
import { Link } from "react-router-dom";
import Select from 'react-select';
import WindowedSelect, { createFilter } from "react-windowed-select";
import CustomOption from './CustomOption';
import Context from '../Context';
import './Header.css'

  export default class Header extends Component {
  
    constructor(props) {

      super(props);
      this.state = {
      };
  
    }

    static contextType = Context; 

    resetApp = () => {

      // If the user switches from desktop to mobile view, say by
      // expanding the dev tools, and then tries to click the home
      // button, the app crashes because the value of the calendar
      // ref gets nulled when the mobile calendar mounts. Have not
      // found a way to update that value, so until I can, reload achieves
      // the same effect without crashing
      if (this.context.calendar.current === null) {
        window.location.reload()
      }

      // This gives us access to methods via the calendar API by creating
      // a ref for each calendar so we can actively manipulate the date
      let calendarApi = this.context.calendar.current.getApi()
      // Set the calendar to today
      calendarApi.today();

      // Update history
      this.context.setLastVisited('/')

      // Get rid of any filters
      this.context.setFilters([])

      // Uncheck any checked checkboxes
      this.context.checkboxes.forEach(checkbox => {
        if (checkbox.state.checked === true) {
          checkbox.setChecked(false)
        }
      })

      // And restore calendar events to their initial state
      this.context.setDefaultReleases(this.context.games)

    }

    // Implements expected keyboard behavior in searchable select
    handleKeyDown = (e) => {
      // No default below turns off esLint rule
      // about default case
      switch(e.key){
           case "Home": e.preventDefault();
               if(e.shiftKey) e.target.selectionStart = 0;
               else e.target.setSelectionRange(0,0);
               break;
           case "End": e.preventDefault();
               const len = e.target.value.length;
               if(e.shiftKey) e.target.selectionEnd = len;
               else e.target.setSelectionRange(len,len);
               break;
          // no default
       }

    };
  
    render() {

      // Pointer for our select
      const customStyles = {
        option: (styles) => ({
          ...styles,
          cursor: 'pointer',
        }),
        control: (styles) => ({
          ...styles,
          cursor: 'pointer',
        }),
      }
  
      return (
        <header className="header">
          <Link to="/" onClick={() => this.resetApp()}><h1 className="logotype">vgCal</h1></Link>
          <WindowedSelect
            styles={customStyles}
            className="search"
            // menuIsOpen={true}
            classNamePrefix="react-select"
            onChange={this.context.handleSelect}
            placeholder="Select a game or type to search for one"
            // Solution to issue of input lag in
            // React Select with large datasets
            filterOption={createFilter({ignoreAccents: false})}
            isClearable={true}
            onKeyDown={this.handleKeyDown}
            components={
              {
                // Solution to problem of poor performance
                // on hover with large datasets. See: https://github.com/JedWatson/react-select/issues/3128#issuecomment-521242192
                Option: CustomOption
              }
            }
            options={this.context.games.map(game => {
              
              const tempObj = {}

              tempObj.value = {
                platforms: game.platforms,
                image: game.boxart_url,
                gameUrl: `/game/${game.game_name.split(' ').join('-')}`,
                gameTitle: game.game_name,
                releaseDate: game.release_date_utc,
                description: game.game_description
              };

              tempObj.label = game.game_name;

              return tempObj;

            }).sort((a, b) => {
                const nameA = a.label;
                const nameB = b.label;
                if (nameA < nameB) {
                  return -1;
                }
                if (nameA > nameB) {
                  return 1;
                }

                return 0;
            })}
          />
          <nav className="header-nav">
            <i onClick={() => this.context.setVisited(false)} className="fas fa-question-circle"></i>
            <a href="https://github.com/bradbautista/" target="_blank" rel="noopener noreferrer"><i className="fab fa-github"></i>
            </a>
            <a href="https://www.linkedin.com/in/bradford-bautista/" target="_blank" rel="noopener noreferrer">
              <i className="fab fa-linkedin-in"></i>
            </a>
          </nav>
          
        </header>
      );
    }
  }