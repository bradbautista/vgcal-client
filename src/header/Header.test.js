import React from 'react';
import ReactDOM from 'react-dom';
import Header from './Header';
import Context from '../Context';
import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';

const someFunc = () => { console.log('Do stuff') };

const games = [{"id":6990,"boxart_url":"https://giantbomb1.cbsistatic.com/uploads/original/16/164924/3177373-0778025400-mediu.png","game_name":"Namcot Collection","game_description":"A collection of NES titles from Namco, from under their Namcot banner.","platforms":"Nintendo Switch","release_date_utc":"June 18, 2020","release_date_iso":"2020-06-18","release_day":"18","release_month":"06","release_year":"2020","release_quarter":null}]

it('Renders the header without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(
    <BrowserRouter>
        <Context.Provider value={ { handleSelect: () => { someFunc() }, games: games, setVisited: () => { someFunc() } } }>
            <Header />
        </Context.Provider>
    </BrowserRouter>, div);
  ReactDOM.unmountComponentAtNode(div);
});

it('Renders the header as expected', () => {
    
    const tree = renderer
        .create(
            
            <BrowserRouter>
                <Context.Provider value={ { handleSelect: () => { someFunc() }, games: games, setVisited: () => { someFunc() } } }>
                    <Header/>
                </Context.Provider>
            </BrowserRouter>
            )
        .toJSON();
    expect(tree).toMatchSnapshot();
        
})