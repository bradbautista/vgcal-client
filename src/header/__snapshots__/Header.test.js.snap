// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders the header as expected 1`] = `
<header
  className="header"
>
  <a
    href="/"
    onClick={[Function]}
  >
    <h1
      className="logotype"
    >
      vgCal
    </h1>
  </a>
  <div
    className="search css-2b097c-container"
    onKeyDown={[Function]}
  >
    <div
      className="react-select__control css-e91lkq-control"
      onMouseDown={[Function]}
      onTouchEnd={[Function]}
    >
      <div
        className="react-select__value-container css-1hwfws3"
      >
        <div
          className="react-select__placeholder css-1wa3eu0-placeholder"
        >
          Select a game or type to search for one
        </div>
        <div
          className="css-xqydr5-Input"
        >
          <div
            className="react-select__input"
            style={
              Object {
                "display": "inline-block",
              }
            }
          >
            <input
              aria-autocomplete="list"
              autoCapitalize="none"
              autoComplete="off"
              autoCorrect="off"
              disabled={false}
              id="react-select-3-input"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              spellCheck="false"
              style={
                Object {
                  "background": 0,
                  "border": 0,
                  "boxSizing": "content-box",
                  "color": "inherit",
                  "fontSize": "inherit",
                  "label": "input",
                  "opacity": 1,
                  "outline": 0,
                  "padding": 0,
                  "width": "1px",
                }
              }
              tabIndex="0"
              type="text"
              value=""
            />
            <div
              style={
                Object {
                  "height": 0,
                  "left": 0,
                  "overflow": "scroll",
                  "position": "absolute",
                  "top": 0,
                  "visibility": "hidden",
                  "whiteSpace": "pre",
                }
              }
            >
              
            </div>
          </div>
        </div>
      </div>
      <div
        className="react-select__indicators css-1hb7zxy-IndicatorsContainer"
      >
        <span
          className="react-select__indicator-separator css-1okebmr-indicatorSeparator"
        />
        <div
          aria-hidden="true"
          className="react-select__indicator react-select__dropdown-indicator css-tlfecz-indicatorContainer"
          onMouseDown={[Function]}
          onTouchEnd={[Function]}
        >
          <svg
            aria-hidden="true"
            className="css-6q0nyr-Svg"
            focusable="false"
            height={20}
            viewBox="0 0 20 20"
            width={20}
          >
            <path
              d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
  <nav
    className="header-nav"
  >
    <i
      className="fas fa-question-circle"
      onClick={[Function]}
    />
    <a
      href="https://github.com/bradbautista/"
      rel="noopener noreferrer"
      target="_blank"
    >
      <i
        className="fab fa-github"
      />
    </a>
    <a
      href="https://www.linkedin.com/in/bradford-bautista/"
      rel="noopener noreferrer"
      target="_blank"
    >
      <i
        className="fab fa-linkedin-in"
      />
    </a>
  </nav>
</header>
`;
